### PR TITLE
feat(connector): [Braintree] implement 3DS card payment for braintree

### PR DESCRIPTION
**Pull request by AkshayaFoiger**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## How did you test it?

Tested It through postman
Screenshot 2023-09-06 at 5 48 46 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/braintree.rs [hunk 1]
```diff
@@ -1248,3 +1251,147 @@ impl api::IncomingWebhook for Braintree {
...
+                .build(),
+        ))
+    }
+    fn handle_response(
+        &self,
+        data: &types::PaymentsCompleteAuthorizeRouterData,
+        res: types::Response,
+    ) -> CustomResult<types::PaymentsCompleteAuthorizeRouterData, errors::ConnectorError> {
+        let connector_api_version = &data.connector_api_version;
+        match self.is_braintree_graphql_version(connector_api_version) {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/braintree.rs:1357]: This check is not required for handle_response, If it is older version error will be thrown in build_request itself

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 2]
```diff
@@ -56,29 +65,48 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for BraintreePaymentsRequest {
...
+                                input: InputClientTokenData {
+                                    client_token: ClientTokenInput {
+                                        merchant_account_id: metadata.merchant_account_id.ok_or(
+                                            errors::ConnectorError::MissingRequiredField {
+                                                field_name: "merchant_account_id",
+                                            },
+                                        )?,
                                     },
-                                )?,
+                                },
```
Comment (ArjunKarthik) [inline crates/router/src/connector/braintree/braintree_graphql_transformers.rs:81]: Move this to try_from()

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 3]
```diff
@@ -56,29 +65,48 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for BraintreePaymentsRequest {
...
+                                    )?,
+                                    merchant_account_id: metadata.merchant_account_id.ok_or(
+                                        errors::ConnectorError::MissingRequiredField {
+                                            field_name: "merchant_account_id",
+                                        },
+                                    )?,
+                                },
+                            },
+                        },
+                    }))
```
Comment (ArjunKarthik) [inline crates/router/src/connector/braintree/braintree_graphql_transformers.rs:108]: Move this to try_from()

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 4]
```diff
@@ -101,6 +129,81 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for BraintreePaymentsRequest {
...
+            | Some(api_models::payments::PaymentMethodData::BankDebit(_))
+            | Some(api_models::payments::PaymentMethodData::BankTransfer(_))
+            | Some(api_models::payments::PaymentMethodData::Crypto(_))
+            | Some(api_models::payments::PaymentMethodData::MandatePayment)
+            | Some(api_models::payments::PaymentMethodData::Reward)
+            | Some(api_models::payments::PaymentMethodData::Upi(_))
+            | Some(api_models::payments::PaymentMethodData::Voucher(_))
+            | Some(api_models::payments::PaymentMethodData::GiftCard(_))
+            | None => Err(errors::ConnectorError::NotImplemented(
+                utils::get_unimplemented_payment_method_error_message("braintree"),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/braintree/braintree_graphql_transformers.rs:200]: It should NotSupported and the message should be "complete authorize flow"

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 5]
```diff
@@ -159,6 +274,47 @@ impl<F, T>
...
+                        },
+                        amount: item.data.request.amount,
+                    }),
+                    mandate_reference: None,
+                    connector_metadata: None,
+                    network_txn_id: None,
+                    connector_response_reference_id: None,
+                }),
+                ..item.data
+            }),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/braintree/braintree_graphql_transformers.rs:317]: Move this to try from

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 6]
```diff
@@ -315,6 +477,141 @@ impl<F, T>
...
```
Comment (ArjunKarthik) [inline crates/router/src/connector/braintree/braintree_graphql_transformers.rs:520]: Move this to try_from and use in BraintreePaymentsResponse and BraintreeAuthResponse try_from conversion

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 7]
```diff
@@ -315,6 +477,141 @@ impl<F, T>
...
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(
+        item: types::ResponseRouterData<
+            F,
+            BraintreePaymentsResponse,
+            types::CompleteAuthorizeData,
+            types::PaymentsResponseData,
+        >,
+    ) -> Result<Self, Self::Error> {
+        match item.response {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/braintree/braintree_graphql_transformers.rs:544]: Create new enum for CompleteAuthorize response

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 8]
```diff
@@ -315,6 +477,141 @@ impl<F, T>
...
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(
+        item: types::ResponseRouterData<
+            F,
+            BraintreeAuthResponse,
+            types::CompleteAuthorizeData,
+            types::PaymentsResponseData,
+        >,
+    ) -> Result<Self, Self::Error> {
+        match item.response {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/braintree/braintree_graphql_transformers.rs:591]: Same here, Create new enum for CompleteAuthorize response

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by prasunna09 — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by prasunna09 — APPROVED** [review]

File: crates/router/src/services/api.rs [hunk 9]
```diff
@@ -1147,6 +1153,92 @@ pub fn build_redirection_form(
...
+                            })
+                            </script>
+                            "#))
+
+
+                        h3 style="text-align: center;" { "Please wait while we process your payment..." }
+                    }
+
+                    (PreEscaped(format!("<script>
```
Comment (SanchithHegde) [inline crates/router/src/services/api.rs:1194]: You can use raw string literals to improve readability.

Comment (AkshayaFoiger) [reply crates/router/src/services/api.rs:1194]: Will take up this change in a different PR

**Pull request merged**